Add EM introducing constants, global variables, and EM Setting variables (#17)
ctsit/redcap_external_module_development_guide: docs/guide_for_devs.md

require_once APP_PATH_DOCROOT . 'ProjectGeneral/footer.php';
```

Don't forget to update docs when you add new features!  
`README.md`
```markdown
Provides pages that say "Hello, world" in the control center and in projects.
```

</details>
<br />

---

### [Accessing Variables]({{ site.repo_root }}exercises/accessing_variables/)

> While working on this module, you will learn how to access constants and variables defined by REDCap. You will also cover using `project-settings` to allow users to set variables and accessing those variables.

Reviewer: ```suggestion
While working on this module, you will learn how to access constants and variables defined by REDCap. You will also use `project-settings` to allow set and access variables.
```

You may display this via a hook or a project plugin page.

The goal of this exercise is to create a module that displays a user's:
1. Username
1. Admin (aka superuser) status
1. Their user rights
1. The current page path
1. The current project's `project_id`
1. The value of a variable set in the module's configuration menu

> Check the [official External Module documentation on functions](https://github.com/vanderbilt/redcap-external-modules/blob/testing/docs/framework/intro.md) for functions which will provide this information for you. Useful phrases to search for are "User" and "projectSetting".

Reviewer: ```suggestion
Check the [official External Module documentation on functions](https://github.com/vanderbilt/redcap-external-modules/blob/testing/docs/framework/intro.md) for functions which provide this information for you. Useful phrases to search for are "User" and "projectSetting".
```

Note that there is an unlisted property of the [`User` object](https://github.com/vanderbilt/redcap-external-modules/blob/testing/docs/framework/intro.md#user-method) that you will need: 

```php
$User = $this->framework->getUser();
$User->username;
```

> If your programming environment has a debugger available, it is highly recommended to use it to see the accessible constants. If you do not have a debugger or cannot connect it to your REDCap docker container, please view the list below. Note that most of these are available as functions.  

Reviewer: ```suggestion
If your programming environment has a debugger available, you can use it to see the accessible constants. If you do not have a debugger or cannot connect it to your REDCap docker container, please view the list below. Note that most of these are available as functions.  
```

<details>
<summary>Useful global constants
</summary>

> - **PAGE**: The current loaded file and its path with the root being the active REDCap folder, (root being, i.e. `redcap_v9.3.5/`); this is the only constant listed that is not easily determined by a function

Reviewer: ```suggestion
- **PAGE**: The path portion of the URL for the current web page. This is the only constant listed that is not easily determined by a function.
```

- **PROJECT_ID**: The numerical ID of the project being viewed
- **USERID**: The username of the user viewing the page
- **SUPER_USER**: A boolean integer indication if the user from **USERID** has admin privileges

</details>
<br />

As before, documentation _must_ be updated when introducing features. This is even more important when adding functionality that users interact with (e.g. an entry in `project-settings` in the `config.json`)!

<details>
<summary>Example Solution Via a Hook
</summary>

> Note that this is a _bare minimum_ implementation; because the goal of this exercise is to find out how to access the variables needed, this solution only aims to show how to access those variables. When displaying information to users, you should _not_ use basic functions like `print_r` (unless the contents will be showing up in the body of a plugin page).  

Reviewer: ```suggestion
This is a _bare minimum_ implementation that demonstrates how to access REDCap variables. The output is _ugly_. An attractive display of those variables is left as an exercise for the student.  
```

Author: My zealous fervor for semicolons is balanced by the use of "left as an exercise to the student".
